### PR TITLE
sql: Added infomationSchmaColumnPrivilege

### PR DESCRIPTION
**Pull request by sum12**
The default value of `granter` and `is_grantable` currently is NULL, 
I picked this from other table.

Also the [postgres documentation](https://www.postgresql.org/docs/8.3/static/infoschema-column-privileges.html) mentions about a `REFERENCES` privilege, However I am assuming its not yet part of cockroachdb

refers #8675

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
Thanks @sum12! This could use some testing though. Check out `pkg/sql/logictest/testdata/logic_test/information_schema`. You can run the logic tests using `make testlogic`.

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/information_schema.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L141Hzd36SB3zyROgYo:-L141Hzd36SB3zyROgYp:b33lwzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/information_schema.go#L50)):*
> ```Go
> 		informationSchemaViewsTable,
> 		informationSchemaUserPrivileges,
> 		informationSchemaColumnPrivileges,
> ```

Keep this list sorted. So this should go first.

---

*[pkg/sql/information_schema.go, line 929 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L141X48T4dX2308q3A-:-L141X48T4dX2308q3A0:b-vjd2s9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/information_schema.go#L929)):*
> ```Go
> }
> 
> var informationSchemaColumnPrivileges = virtualSchemaTable{
> ```

Move this above `informationSchemaColumnsTable`.

---

*[pkg/sql/information_schema.go, line 945 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L143I0hgdVAPZDLxmG0:-L143I0icRoyApMC-HTP:bwuvp1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/information_schema.go#L945)):*
> ```Go
> 		return forEachTableDesc(ctx, p, prefix, func(db *sqlbase.DatabaseDescriptor, table *sqlbase.TableDescriptor) error {
> 			for _, u := range table.Privileges.Show() {
> 				for _, tbprivilege := range u.Privileges {
> ```

`s/tbprivilege/p/` and also `s/privilege/p/` for `informationSchemaTablePrivileges`.

---

*[pkg/sql/information_schema.go, line 951 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L143703rn6TIivGvAve:-L143703rn6TIivGvAvf:b-7ijqj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/information_schema.go#L951)):*
> // table_catalog,

 no comma here. Could you remove it from `informationSchemaTablePrivileges` as well?

---

*[pkg/sql/privilege/privilege.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L143lvEUDdMYJD0kMZ8:-L143lvFxB8MBAudGHgU:b-moymv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/privilege/privilege.go#L48)):*
> ```Go
> 	ReadData        = List{GRANT, SELECT}
> 	ReadWriteData   = List{GRANT, SELECT, INSERT, DELETE, UPDATE}
> 	ColumnPrivilege = List{SELECT, INSERT, UPDATE}
> ```

Give this a comment about how these are the set of privileges applicable to a column. Also, rename to `ColumnData`.

---

*[pkg/sql/privilege/privilege.go, line 142 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L144at-dvsh5jnDslHO:-L144at-dvsh5jnDslHP:bdqljv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fc2fd636fc1a40785f2dc009b4292362c562fa6/pkg/sql/privilege/privilege.go#L142)):*
> ```Go
> }
> 
> func (ls List) AnyIn(m uint32) bool {
> ```

This could use a comment and should probably be renamed to `Contains`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-:-L144-0p9_A6h3I3S90v:bqsnizy)*

**Comment by sum12** [comment]
@nvanbenschoten thanks for the review.
I have changed the approach and I think its a bit more clean than before.
And force pushed my changes, hope it doesnt get confusing.

**Comment by sum12** [comment]
@nvanbenschoten thanks for pointing me to test cases,
They have been corrected and `make testlogic` passed.
If the implementation is fine I will tests

File: pkg/sql/information_schema.go [hunk 1]
```diff
@@ -507,14 +548,14 @@ CREATE TABLE information_schema.table_privileges (
 	populate: func(ctx context.Context, p *planner, prefix string, addRow func(...tree.Datum) error) error {
 		return forEachTableDesc(ctx, p, prefix, func(db *sqlbase.DatabaseDescriptor, table *sqlbase.TableDescriptor) error {
 			for _, u := range table.Privileges.Show() {
-				for _, privilege := range u.Privileges {
+				for _, p := range u.Privileges {
```
Comment (mberhault) [inline pkg/sql/information_schema.go]: same as above, no need to shadow `p`.

File: pkg/sql/information_schema.go [hunk 2]
```diff
@@ -98,6 +99,46 @@ func dIntFnOrNull(fn func() (int32, bool)) tree.Datum {
...
+	TABLE_NAME STRING NOT NULL DEFAULT '',
+	COLUMN_NAME STRING NOT NULL DEFAULT '',
+	PRIVILEGE_TYPE STRING NOT NULL DEFAULT '',
+	IS_GRANTABLE BOOL NOT NULL DEFAULT FALSE
+);
+`,
+	populate: func(ctx context.Context, p *planner, prefix string, addRow func(...tree.Datum) error) error {
+		return forEachTableDesc(ctx, p, prefix, func(db *sqlbase.DatabaseDescriptor, table *sqlbase.TableDescriptor) error {
+			for _, u := range table.Privileges.Users {
+				for _, p := range privilege.ColumnData {
```
Comment (mberhault) [inline pkg/sql/information_schema.go]: `p *planner` already exists. Let's do the same as below and use `privilege`

File: pkg/sql/logictest/testdata/logic_test/information_schema [hunk 3]
```diff
@@ -107,6 +107,7 @@ test
 query T
 SHOW TABLES FROM information_schema
 ----
+column_privileges
```
Comment (mberhault) [inline pkg/sql/logictest/testdata/logic_test/information_schema]: can you add a `show columns` for the new table, and perhaps even a `select *`

File: pkg/sql/privilege/privilege.go [hunk 4]
```diff
@@ -45,6 +45,7 @@ const (
 var (
 	ReadData      = List{GRANT, SELECT}
 	ReadWriteData = List{GRANT, SELECT, INSERT, DELETE, UPDATE}
+	ColumnData    = List{SELECT, INSERT, UPDATE} // privileges for column level granularity
```
Comment (mberhault) [inline pkg/sql/privilege/privilege.go]: This should be in `information_schema` as it's only used there.

**Comment by nvanbenschoten** [comment]
This should be good to go once @mberhault and my remaining comments are addressed. Thanks for the help here! 

---

Reviewed 1 of 3 files at r3, 1 of 2 files at r5.
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/information_schema.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L14-w3AZl_fktJREPQf-r3-118:-L1OG67-fQvQxEknI77W:bqbhqfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/information_schema.go#L118)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

`p *planner` already exists. Let's do the same as below and use `privilege`
</blockquote></details>

Sorry, I said to do this down below but didn't realize we'd be shadowing `p`. I was trying to avoid shadowing the `privilege` package.  @sum12 how about `priv`?

---

*[pkg/sql/information_schema.go, line 556 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGvJ11HRP-afr2M0Q:-L1OGvJ11HRP-afr2M0R:bilduth) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/information_schema.go#L556)):*
> ```Go
> 						tree.DNull,                  // grantor
> 						tree.NewDString(u.User),     // grantee
> 						defString,                   // table_catalog,
> ```

`s/table_catalog,/table_catalog/`

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1095 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGaZ9Hc2Pu0JC2iMA:-L1OGaZ9Hc2Pu0JC2iMB:beyzit1) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1095)):*
> infomration_schama

information_schema

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1099 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGeHxhyxzHZl9AqCx:-L1OGeHxhyxzHZl9AqCy:b-bw8g46) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1099)):*
> ```
> SHOW COLUMNS FROM information_schema.column_privileges
> ----
> Field           Type    Null   Default      Indices  
> ```

Let's remove the extra spaces at the end of these lines.

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1110 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGky9zc7kpunER-PP:-L1OGky9zc7kpunER-PQ:b-mykiue) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1110)):*
> ```
> 
> 
> query TTTTTTTT
> ```

Add `colnames` here.

---

*[pkg/sql/privilege/privilege.go, line 48 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L14-w3AZl_fktJREPQg-r3-48:-L1OGpn4bjCVg_kLgFr2:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/privilege/privilege.go#L48)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This should be in `information_schema` as it's only used there.
</blockquote></details>

+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-:-L1OH2HgyKTy3NNSlZkH:b-kdexed)*

**Comment by sum12** [comment]
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/information_schema.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L14-w3AZl_fktJREPQf-r3-118:-L1S2DWEV8vChFev9YcW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/information_schema.go#L118)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Sorry, I said to do this down below but didn't realize we'd be shadowing `p`. I was trying to avoid shadowing the `privilege` package.  @sum12 how about `priv`?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 110 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L170ixW3d3_zI7J_kL8-r3-110:-L1S2EwUr2FppBiZFDTx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/logictest/testdata/logic_test/information_schema#L110)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

can you add a `show columns` for the new table, and perhaps even a `select *`
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1095 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGaZ9Hc2Pu0JC2iMA:-L1S2FTncLyGbe-YNaHu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1095)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> infomration_schama

information_schema
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1099 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGeHxhyxzHZl9AqCx:-L1S1w-NfISYh_s--wFK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1099)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's remove the extra spaces at the end of these lines.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1110 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1OGky9zc7kpunER-PP:-L1S1qmKk4kiieyolA_G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/673af65128da2434ea5a9b8fc94747b991efa2b5/pkg/sql/logictest/testdata/logic_test/information_schema#L1110)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add `colnames` here.
</blockquote></details>

Done.

---

*[pkg/sql/privilege/privilege.go, line 48 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L14-w3AZl_fktJREPQg-r3-48:-L1S35yKER6R49KP7uWC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/privilege/privilege.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031)*

**Comment by nvanbenschoten** [comment]
Two more minor changes then all the commits can be squashed down to one and rebased off of `master`.

---

Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/information_schema.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L14-w3AZl_fktJREPQf-r3-118:-L1TyFJt_DNwBSI15IUA:b610gzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e2aacb8803eb9df50d6fcf8d5bf68c697900687/pkg/sql/information_schema.go#L118)):*
<details><summary><i>Previously, sum12 (Sumit) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`priv` here as well:

```
for _, priv := range columndata {
```

---

*[pkg/sql/logictest/testdata/logic_test/information_schema, line 1111 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-L1TymHQLTS5_0MQzUCr:-L1TymHQLTS5_0MQzUCs:b-o4g2qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da1bdcb295c9e5d4e22ad71a93401076aa09cd9/pkg/sql/logictest/testdata/logic_test/information_schema#L1111)):*
> ```
> 
> query TTTTTTTT colnames
> SELECT * FROM information_schema.column_privileges WHERE table_name = 'eventlog'
> ```

nit: Could you make this `ORDER BY column_name`?

Also, do the tests pass right now? It doesn't look like `colnames` had an effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21031#-:-L1Tyz3pXlY4kN5Djpia:bgh79te)*

**Comment by sjamgade** [comment]
The tests should now succeed.
once the results are updated, will squash and rebase

**Comment by sum12** [comment]
the test should now succeed.
Once the result is here, will then squash and rebase

**Comment by nvanbenschoten** [comment]
@sum12 it looks like this has a conflict with another change. Could you rebase off master again and squash all the commits into one? Then we can wait for CI and merge immediately.

**Comment by sum12** [comment]
@nvanbenschoten, tests passed, this can be merged now.

**Pull request merged**